naming/pubsub.md

```diff
@@ -0,0 +1,95 @@
+  - [Protocol](#protocol)
+  - [Overview](#overview)
+  - [API Spec](#api-spec)
+  - [Integration with IPFS](#integration-with-ipfs)
+
+# Introduction
+
+Each time a node publishes an updated IPNS record for a particular key it is propagated by the router into the network where network nodes can choose to accept or reject the new record.
+When a node attempts to retrieve an IPNS record from the network it uses the router to query for the IPNS record(s) associated with the IPNS key; the node then validates the received records.
+
+In this spec we address building a router based on a PubSub system, particularly focusing on libp2p PubSub.
+
+# PubSub Protocol Overview
+
+The protocol has four components:
+- [IPNS Records and Validation](/README.md)
+- [libp2p PubSub](https://github.com/libp2p/specs/tree/master/pubsub)
+- Translating an IPNS record name to/from a PubSub topic
+- Layering persistence onto libp2p PubSub
+
+# Translating an IPNS record name to/from a PubSub topic
+
+For a given IPNS local record key described in the IPNS Specification the PubSub topic is:
+
+**Topic format:** `/record/base64url-unpadded(key)`
+
+where base64url-unpadded is an unpadded base64url as specified in [IETF RFC 4648](https://tools.ietf.org/html/rfc4648)
+
+# Layering persistence onto libp2p PubSub
+
+libp2p PubSub does not have any notion of persistent data built into it. However, we can layer persistence on top of PubSub by utilizing [libp2p Fetch](https://github.com/libp2p/specs/).
+
+The protocol has the following steps:
+1. Start State: Node `A` subscribes to the PubSub topic `t` corresponding to the local IPNS record key `k`
+2. `A` notices that a node `B` has connected to it and subscribed to `t`
+3. Some time passes (might be 0 seconds, or could use a more complex system to determine the duration)
+4. `A` sends `B` a Fetch request for `k`
+5. If Fetch returns a record that supersedes `A`'s current record then `A` updates its record and Publishes it to the network
+
+Note: PubSub does not guarantee that the a message sent by a peer `A` will be received by a peer `B` and it's possible
+(e.g. in systems like [gossipsub](https://github.com/libp2p/specs/tree/master/pubsub/gossipsub))
+that this is true even if `A` and `B` are already connected. Therefore, whenever `A` notices **any** node that has
+connected to it and subscribed to `t` it should run the Fetch protocol as described above. However, developers may have routers
+with properties that allow the amount of time in step 3 to increase arbitrarily large (including infinite) amounts.
+
+# Protocol
+
+A node `A` putting and getting updates to an IPNS key `k`, with computed PubSub topic `t`
+
+1. PubSub subscribe to `t`
+2. Run the persistence protocol, both to fetch data and return data to those that request it
+3. When updating a record do a PubSub Publish and keep the record locally
+4. When receiving a record if it's better than the current record keep it and republish the message

```

Comment:
Maybe, I figured "best" was something that should be defined in the IPNS Spec and I tried to stay away from doubly defining things.

Thoughts?

Reply:
yes, "best" is ok, as long we define it before this list.

Reply:
I added defining "best" to this issue https://github.com/ipfs/specs/issues/205#issuecomment-521615574 since it should be defined for IPNS in general instead of being transport specific.

Reply:
Also I might be wrong here, but I like the more ambiguous language of "better" over "more up to date" because of oddities like if the sequence numbers are the same, but the data fields are different.